<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Wei Liu" />

<meta name="date" content="2024-03-18" />

<title>CoFAST: NSCLC CosMx data coembedding</title>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">CoFAST: NSCLC CosMx data coembedding</h1>
<h4 class="author">Wei Liu</h4>
<h4 class="date">2024-03-18</h4>



<p>This vignette introduces the CoFAST workflow for the analysis of NSCLC CosMx spatial transcriptomics dataset. In this vignette, the workflow of CoFAST consists of three steps</p>
<ul>
<li>Independent preprocessing and model setting</li>
<li>Coembedding dimension reduction</li>
<li>Downstream analysis (i.e.Â , signature gene analysis, visualization of cell types and coembeddings,)</li>
</ul>
<div id="load-and-view-data" class="section level2">
<h2>Load and view data</h2>
<p>We demonstrate the use of CoFAST to NSCLC data, which can be downloaded to the current working path by the following command:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">set.seed</span>(<span class="dv">2024</span>) <span class="co"># set a random seed for reproducibility.</span></a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw">library</span>(ProFAST) <span class="co"># load the package of FAST method</span></a>
<a class="sourceLine" id="cb1-3" title="3"><span class="kw">data</span>(CosMx_subset)</a>
<a class="sourceLine" id="cb1-4" title="4">CosMx_subset</a></code></pre></div>
<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw">library</span>(Seurat)</a></code></pre></div>
</div>
<div id="preprocessing" class="section level2">
<h2>Preprocessing</h2>
<p>First, we normalize the data.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">CosMx_subset &lt;-<span class="st"> </span><span class="kw">NormalizeData</span>(CosMx_subset)</a></code></pre></div>
<p>Then, we select the variable genes.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1">CosMx_subset &lt;-<span class="st"> </span><span class="kw">FindVariableFeatures</span>(CosMx_subset)</a></code></pre></div>
</div>
<div id="coembedding-using-fast" class="section level2">
<h2>Coembedding using FAST</h2>
<p>We introduce how to use FAST to perform coembedding for this CosMx data. First, we determine the dimension of coembeddings. Then, we select the variable genes.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">dat_cor &lt;-<span class="st"> </span><span class="kw">diagnostic.cor.eigs</span>(CosMx_subset)</a>
<a class="sourceLine" id="cb5-2" title="2">q_est &lt;-<span class="st"> </span><span class="kw">attr</span>(dat_cor, <span class="st">&quot;q_est&quot;</span>)</a>
<a class="sourceLine" id="cb5-3" title="3"><span class="kw">cat</span>(<span class="st">&quot;q_est = &quot;</span>, q_est, <span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>)</a></code></pre></div>
<p>Subsequently, we calculate coembeddings by utilizing FAST, and observe that the <code>reductions</code> field acquires an additional component named <code>fast</code>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1">pos &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(CosMx_subset<span class="op">@</span>meta.data[,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>, <span class="st">&quot;y&quot;</span>)]) <span class="co"># Extract the spatial coordinates</span></a>
<a class="sourceLine" id="cb6-2" title="2">Adj_sp &lt;-<span class="st"> </span><span class="kw">AddAdj</span>(pos) <span class="co">## calculate the adjacency matrix</span></a>
<a class="sourceLine" id="cb6-3" title="3">CosMx_subset &lt;-<span class="st"> </span><span class="kw">NCFM_fast</span>(CosMx_subset, <span class="dt">Adj_sp =</span> Adj_sp, <span class="dt">q =</span> q_est)</a>
<a class="sourceLine" id="cb6-4" title="4">CosMx_subset</a></code></pre></div>
</div>
<div id="downstream-analysis" class="section level2">
<h2>Downstream analysis</h2>
<p>In the following, we show how to find the signature genes based on comebeddings. First, we calculate the distance matrix.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">CosMx_subset &lt;-<span class="st"> </span><span class="kw">pdistance</span>(CosMx_subset, <span class="dt">reduction =</span> <span class="st">&quot;fast&quot;</span>)</a></code></pre></div>
<p>Next, we find the signature genes for each cell type</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="kw">print</span>(<span class="kw">table</span>(CosMx_subset<span class="op">$</span>cell_type))</a>
<a class="sourceLine" id="cb8-2" title="2"><span class="kw">Idents</span>(CosMx_subset) &lt;-<span class="st"> </span>CosMx_subset<span class="op">$</span>cell_type</a>
<a class="sourceLine" id="cb8-3" title="3">df_sig_list &lt;-<span class="st"> </span><span class="kw">find.signature.genes</span>(CosMx_subset)</a>
<a class="sourceLine" id="cb8-4" title="4"><span class="kw">str</span>(df_sig_list)</a></code></pre></div>
<p>Then, we obtain the top five signature genes and organize them into a data.frame. Next, we calculate the UMAP projections of coembeddings. The colname <code>distance</code> means the distance between gene (i.e., MS4A1) and cells with the specific cell type (i.e., B cell), which is calculated based on the coembedding of genes and cells in the coembedding space. The distance is smaller, the association between gene and the cell type is stronger. The colname <code>expr.prop</code> represents the expression proportion of the gene (i.e., MS4A1) within the cell type (i.e., B cell). The colname <code>label</code> means the cell types and colname <code>gene</code> denotes the gene name. By the data.frame object, we know <code>MS4A1</code> is the one of the top signature gene of B cell.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">dat &lt;-<span class="st"> </span><span class="kw">get.top.signature.dat</span>(df_sig_list, <span class="dt">ntop =</span> <span class="dv">2</span>, <span class="dt">expr.prop.cutoff =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb9-2" title="2"><span class="kw">head</span>(dat)</a></code></pre></div>
<p>Next, we calculate the UMAP projections of coembeddings of cells and the selected signature genes.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">CosMx_subset &lt;-<span class="st"> </span><span class="kw">coembedding_umap</span>(</a>
<a class="sourceLine" id="cb10-2" title="2">  CosMx_subset, <span class="dt">reduction =</span> <span class="st">&quot;fast&quot;</span>, <span class="dt">reduction.name =</span> <span class="st">&quot;UMAP&quot;</span>,</a>
<a class="sourceLine" id="cb10-3" title="3">  <span class="dt">gene.set =</span> <span class="kw">unique</span>(dat<span class="op">$</span>gene))</a></code></pre></div>
<p>Furthermore, we visualize the cells and top two signature genes of tumor 5 in the UMAP space of coembedding. We observe that the UMAP projections of the two signature genes are near to B cells, which indicates these genes are enriched in B cells.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="co">## choose beutifual colors</span></a>
<a class="sourceLine" id="cb11-2" title="2">cols_cluster &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;black&quot;</span>, PRECAST<span class="op">::</span><span class="kw">chooseColors</span>(<span class="dt">palettes_name =</span> <span class="st">&quot;Blink 23&quot;</span>, <span class="dt">n_colors =</span> <span class="dv">21</span>, <span class="dt">plot_colors =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb11-3" title="3">p1 &lt;-<span class="st"> </span><span class="kw">coembed_plot</span>(</a>
<a class="sourceLine" id="cb11-4" title="4">   CosMx_subset, <span class="dt">reduction =</span> <span class="st">&quot;UMAP&quot;</span>,</a>
<a class="sourceLine" id="cb11-5" title="5">   <span class="dt">gene_txtdata =</span> <span class="kw">subset</span>(dat, label<span class="op">==</span><span class="st">&#39;tumor 5&#39;</span>),</a>
<a class="sourceLine" id="cb11-6" title="6">   <span class="dt">cols=</span>cols_cluster, <span class="dt">pt_text_size =</span> <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb11-7" title="7">p1</a></code></pre></div>
<p>Then, we visualize the cells and top two signature genes of all involved cell types in the UMAP space of coembedding. We observe that the UMAP projections of the signature genes are near to the corresponding cell type, which indicates these genes are enriched in the corresponding cells.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">p2 &lt;-<span class="st"> </span><span class="kw">coembed_plot</span>(</a>
<a class="sourceLine" id="cb12-2" title="2">   CosMx_subset, <span class="dt">reduction =</span> <span class="st">&quot;UMAP&quot;</span>,</a>
<a class="sourceLine" id="cb12-3" title="3">   <span class="dt">gene_txtdata =</span> dat, <span class="dt">cols=</span>cols_cluster, </a>
<a class="sourceLine" id="cb12-4" title="4">   <span class="dt">pt_text_size =</span> <span class="dv">3</span>, <span class="dt">alpha=</span><span class="fl">0.2</span>)</a>
<a class="sourceLine" id="cb12-5" title="5">p2</a></code></pre></div>
<p>In addtion, we can fully take advantages of the visualization functions in <code>Seurat</code> package for visualization. The following is an example that visualizes the cell types on the UMAP space.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1">cols_type &lt;-<span class="st"> </span>cols_cluster[<span class="op">-</span><span class="dv">1</span>]</a>
<a class="sourceLine" id="cb13-2" title="2"><span class="kw">names</span>(cols_type)&lt;-<span class="st">  </span><span class="kw">sort</span>(<span class="kw">levels</span>(<span class="kw">Idents</span>(CosMx_subset)))</a>
<a class="sourceLine" id="cb13-3" title="3"><span class="kw">DimPlot</span>(CosMx_subset, <span class="dt">reduction =</span> <span class="st">&#39;UMAP&#39;</span>, <span class="dt">cols=</span>cols_type)</a></code></pre></div>
<p>Then, there is another example that we plot the first two signature genes of Tumor 5 on UMAP space, in which we observed the high expression in B cells in constrast to other cell types.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1"><span class="kw">FeaturePlot</span>(CosMx_subset, <span class="dt">reduction =</span> <span class="st">&#39;UMAP&#39;</span>, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;PSCA&quot;</span>, <span class="st">&quot;CEACAM6&quot;</span>))</a></code></pre></div>
<details>
<p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1"><span class="kw">sessionInfo</span>()</a>
<a class="sourceLine" id="cb15-2" title="2"><span class="co">#&gt; R version 4.2.1 (2022-06-23 ucrt)</span></a>
<a class="sourceLine" id="cb15-3" title="3"><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span></a>
<a class="sourceLine" id="cb15-4" title="4"><span class="co">#&gt; Running under: Windows 10 x64 (build 22621)</span></a>
<a class="sourceLine" id="cb15-5" title="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb15-6" title="6"><span class="co">#&gt; Matrix products: default</span></a>
<a class="sourceLine" id="cb15-7" title="7"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb15-8" title="8"><span class="co">#&gt; locale:</span></a>
<a class="sourceLine" id="cb15-9" title="9"><span class="co">#&gt; [1] LC_COLLATE=C                               </span></a>
<a class="sourceLine" id="cb15-10" title="10"><span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.utf8   </span></a>
<a class="sourceLine" id="cb15-11" title="11"><span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.utf8</span></a>
<a class="sourceLine" id="cb15-12" title="12"><span class="co">#&gt; [4] LC_NUMERIC=C                               </span></a>
<a class="sourceLine" id="cb15-13" title="13"><span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.utf8    </span></a>
<a class="sourceLine" id="cb15-14" title="14"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb15-15" title="15"><span class="co">#&gt; attached base packages:</span></a>
<a class="sourceLine" id="cb15-16" title="16"><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></a>
<a class="sourceLine" id="cb15-17" title="17"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb15-18" title="18"><span class="co">#&gt; loaded via a namespace (and not attached):</span></a>
<a class="sourceLine" id="cb15-19" title="19"><span class="co">#&gt;  [1] digest_0.6.33   R6_2.5.1        jsonlite_1.8.7  evaluate_0.21  </span></a>
<a class="sourceLine" id="cb15-20" title="20"><span class="co">#&gt;  [5] cachem_1.0.8    rlang_1.1.1     cli_3.4.1       rstudioapi_0.14</span></a>
<a class="sourceLine" id="cb15-21" title="21"><span class="co">#&gt;  [9] jquerylib_0.1.4 bslib_0.5.0     rmarkdown_2.23  tools_4.2.1    </span></a>
<a class="sourceLine" id="cb15-22" title="22"><span class="co">#&gt; [13] xfun_0.39       yaml_2.3.7      fastmap_1.1.1   compiler_4.2.1 </span></a>
<a class="sourceLine" id="cb15-23" title="23"><span class="co">#&gt; [17] htmltools_0.5.5 knitr_1.43      sass_0.4.7</span></a></code></pre></div>
</details>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
