<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Wei Liu" />

<meta name="date" content="2025-03-20" />

<title>FAST: two DLPFC sections</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">FAST: two DLPFC sections</h1>
<h4 class="author">Wei Liu</h4>
<h4 class="date">2025-03-20</h4>



<p>This vignette introduces the FAST workflow for the analysis of two
humn dorsolateral prefrontal cortex (DLPFC) spatial transcriptomics
dataset. FAST workflow is based on the <code>PRECASTObj</code> object
created in the <code>PRECAST</code> R package and the workflow of FAST
is similar to that of PRECAST; see the vignette (<a href="https://feiyoung.github.io/PRECAST/articles/PRECAST.BreastCancer.html" class="uri">https://feiyoung.github.io/PRECAST/articles/PRECAST.BreastCancer.html</a>)
for the workflow of PRECAST. In this vignette, the workflow of FAST
consists of three steps</p>
<ul>
<li>Independent preprocessing and model setting</li>
<li>Spatial dimension reduction using FAST</li>
<li>Downstream analysis (i.e.Â , embedding alignment and spatial
clustering using iSC-MEB, visualization of clusters and embeddings,
remove unwanted variation, combined differential expression
analysis)</li>
<li>Except for the above downstream analyses, cell-cell interaction
analysis and trajectory inference can also be performed on the basis of
the embeddings obtained by FAST.</li>
</ul>
<p>We demonstrate the use of FAST to two DLPFC Visium data that are <a href="https://github.com/feiyoung/ProFAST/tree/main/vignettes_data">here</a>,
which can be downloaded to the current working path by the following
command:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>githubURL <span class="ot">&lt;-</span> <span class="st">&quot;https://github.com/feiyoung/FAST/blob/main/vignettes_data/seulist2_ID9_10.RDS?raw=true&quot;</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">download.file</span>(githubURL,<span class="st">&quot;seulist2_ID9_10.RDS&quot;</span>,<span class="at">mode=</span><span class="st">&#39;wb&#39;</span>)</span></code></pre></div>
<p>Then load to R</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>dlpfc2 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;./seulist2_ID9_10.RDS&quot;</span>)</span></code></pre></div>
<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">library</span>(ProFAST) <span class="co"># load the package of FAST method</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">library</span>(PRECAST)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span></code></pre></div>
<div id="view-the-dlpfc-data" class="section level2">
<h2>View the DLPFC data</h2>
<p>First, we view the the two spatial transcriptomics data with Visium
platform. There are ~15000 genes for each data batch and ~7000 spots in
total.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>dlpfc2 <span class="do">## a list including three Seurat object with default assay: RNA</span></span></code></pre></div>
<p>Check the content in <code>dlpfc2</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">head</span>(dlpfc2[[<span class="dv">1</span>]])</span></code></pre></div>
<p>We observed the genes are Ensembl IDs. In the following, we will
transfer the Ensembl IDs to gene symbols for matching the housekeeping
genes in the downstream analysis for removing the unwanted
variations.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">row.names</span>(dlpfc2[[<span class="dv">1</span>]])[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
</div>
<div id="create-a-precastobject-object" class="section level2">
<h2>Create a PRECASTObject object</h2>
<p>We show how to create a PRECASTObject object step by step. First, we
create a Seurat list object using the count matrix and meta data of each
data batch.</p>
<ul>
<li>Note: the spatial coordinates must be contained in the meta data and
named as <code>row</code> and <code>col</code>, which benefits the
identification of spaital coordinates by FAST</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="do">## Get the gene-by-spot read count matrices</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>countList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(dlpfc2, <span class="cf">function</span>(x) x[[<span class="st">&quot;RNA&quot;</span>]]<span class="sc">@</span>counts)</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>M <span class="ot">&lt;-</span> <span class="fu">length</span>(countList)</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="do">### transfer the Ensembl ID to symbol</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="cf">for</span>(r <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>M){</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>  <span class="fu">row.names</span>(countList[[r]]) <span class="ot">&lt;-</span> <span class="fu">transferGeneNames</span>(<span class="fu">row.names</span>(countList[[r]]), <span class="at">now_name =</span> <span class="st">&quot;ensembl&quot;</span>,</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>                                                 <span class="at">to_name=</span><span class="st">&quot;symbol&quot;</span>,</span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>                                                 <span class="at">species=</span><span class="st">&quot;Human&quot;</span>, <span class="at">Method=</span><span class="st">&#39;eg.db&#39;</span>)</span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>}</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="do">## Check the spatial coordinates: Yes, they are named as &quot;row&quot; and &quot;col&quot;!</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">head</span>(dlpfc2[[<span class="dv">1</span>]]<span class="sc">@</span>meta.data))</span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a><span class="do">## Get the meta data of each spot for each data batch</span></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a>metadataList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(dlpfc2, <span class="cf">function</span>(x) x<span class="sc">@</span>meta.data)</span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a></span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a></span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a><span class="do">## ensure the row.names of metadata in metaList are the same as that of colnames count matrix in countList</span></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a><span class="cf">for</span>(r <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>M){</span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a>  <span class="fu">row.names</span>(metadataList[[r]]) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(countList[[r]])</span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a>}</span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a></span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a></span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a><span class="do">## Create the Seurat list  object</span></span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a></span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a>seuList <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb7-27"><a href="#cb7-27" tabindex="-1"></a><span class="cf">for</span>(r <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>M){</span>
<span id="cb7-28"><a href="#cb7-28" tabindex="-1"></a>  seuList[[r]] <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> countList[[r]], <span class="at">meta.data=</span>metadataList[[r]], <span class="at">project =</span> <span class="st">&quot;FASTdlpfc2&quot;</span>)</span>
<span id="cb7-29"><a href="#cb7-29" tabindex="-1"></a>}</span>
<span id="cb7-30"><a href="#cb7-30" tabindex="-1"></a><span class="fu">print</span>(seuList)</span>
<span id="cb7-31"><a href="#cb7-31" tabindex="-1"></a><span class="fu">row.names</span>(seuList[[<span class="dv">1</span>]])[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<div id="prepare-the-precastobject-with-preprocessing-step." class="section level3">
<h3>Prepare the PRECASTObject with preprocessing step.</h3>
<p>Next, we use <code>CreatePRECASTObject()</code> to create a
PRECASTObject object based on the Seurat list object
<code>seuList</code>. Here, we select the top 2000 highly variable genes
for followed analysis. Users are able to see <a href="https://feiyoung.github.io/PRECAST/articles/PRECAST.BreastCancer.html" class="uri">https://feiyoung.github.io/PRECAST/articles/PRECAST.BreastCancer.html</a>
for what is done in this function.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="do">## Create PRECASTObject</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2023</span>)</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">CreatePRECASTObject</span>(seuList, <span class="at">project =</span> <span class="st">&quot;FASTdlpfc2&quot;</span>, <span class="at">gene.number =</span> <span class="dv">2000</span>, <span class="at">selectGenesMethod =</span> <span class="st">&quot;HVGs&quot;</span>,</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>    <span class="at">premin.spots =</span> <span class="dv">20</span>, <span class="at">premin.features =</span> <span class="dv">20</span>, <span class="at">postmin.spots =</span> <span class="dv">1</span>, <span class="at">postmin.features =</span> <span class="dv">10</span>)</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="do">## User can retain the raw seuList by the following commond.</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="do">##  PRECASTObj &lt;-  CreatePRECASTObject(seuList, customGenelist=row.names(seuList[[1]]), rawData.preserve = TRUE)</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="do">## check the number of genes/features after filtering step</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>PRECASTObj<span class="sc">@</span>seulist</span></code></pre></div>
</div>
</div>
<div id="fit-fast-for-this-data" class="section level2">
<h2>Fit FAST for this data</h2>
<div id="add-the-model-setting" class="section level3">
<h3>Add the model setting</h3>
<p>Add adjacency matrix list and parameter setting of FAST. More model
setting parameters can be found in <code>model_set_FAST()</code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="do">## seuList is null since the default value `rawData.preserve` is FALSE.</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>PRECASTObj<span class="sc">@</span>seuList</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="do">## Add adjacency matrix list for a PRECASTObj object to prepare for FAST model fitting.</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">AddAdjList</span>(PRECASTObj, <span class="at">platform =</span> <span class="st">&quot;Visium&quot;</span>)</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="do">## Add a model setting in advance for a PRECASTObj object: verbose =TRUE helps outputing the information in the algorithm; </span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">AddParSettingFAST</span>(PRECASTObj, <span class="at">verbose=</span><span class="cn">TRUE</span>)</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="do">## Check the parameters</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>PRECASTObj<span class="sc">@</span>parameterList</span></code></pre></div>
</div>
<div id="fit-fast" class="section level3">
<h3>Fit FAST</h3>
<p>For function <code>FAST</code>, users can specify the number of
factors <code>q</code> and the fitted model <code>fit.model</code>. The
<code>q</code> sets the number of spatial factors to be extracted, and a
lareger one means more information to be extracted but higher
computaional cost. The <code>fit.model</code> specifies the version of
FAST to be fitted. The Gaussian version (<code>gaussian</code>) models
the log-count matrices while the Poisson verion (<code>poisson</code>)
models the count matrices; default as <code>poisson</code>. For a fast
implementation, we run Gaussian version of FAST here. (Note: The
computational time required to run the analysis on personal PCs is
approximately ~1 minute.)</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="do">### set q= 15 here</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2023</span>)</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">FAST</span>(PRECASTObj, <span class="at">q=</span><span class="dv">15</span>, <span class="at">fit.model=</span><span class="st">&#39;gaussian&#39;</span>)</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="do">### Check the results</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="fu">str</span>(PRECASTObj<span class="sc">@</span>resList)</span></code></pre></div>
<details>
<summary>
<strong>Run possion version</strong>
</summary>
<p>Users can also use the possion version by the following command:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2023</span>)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">FAST</span>(PRECASTObj, <span class="at">q=</span><span class="dv">15</span>)</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="do">### Check the results</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="fu">str</span>(PRECASTObj<span class="sc">@</span>resList)</span></code></pre></div>
</details>
</div>
<div id="evaluate-the-adjusted-mcfaddens-pseudo-r-square" class="section level3">
<h3>Evaluate the adjusted McFaddenâs pseudo R-square</h3>
<p>Next, we investigate the performance of dimension reduction by
calculating the adjusted McFaddenâs pseudo R-square for each data batch.
The manual annotations are regarded as the groud truth in the
<code>meta.data</code> of each component of
<code>PRECASTObj@seulist</code>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="do">## Obtain the true labels</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>yList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(PRECASTObj<span class="sc">@</span>seulist, <span class="cf">function</span>(x) x<span class="sc">$</span>layer_guess_reordered)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="do">### Evaluate the MacR2</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>MacVec <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(PRECASTObj<span class="sc">@</span>seulist), <span class="cf">function</span>(r) <span class="fu">get_r2_mcfadden</span>(PRECASTObj<span class="sc">@</span>resList<span class="sc">$</span>FAST<span class="sc">$</span>hV[[r]], yList[[r]]))</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="do">### output them</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="fu">print</span>(MacVec)</span></code></pre></div>
</div>
</div>
<div id="embedding-alignment-and-spatial-clustering-using-isc-meb" class="section level2">
<h2>Embedding alignment and spatial clustering using iSC-MEB</h2>
<p>Based on the embeddings from FAST, we use <code>iSC-MEB</code> to
jointly align the embeddings and perform clustering. In this downstream
analysis, other methods for embedding alignment and clustering can be
also used. In the vignette of the simulated ST data, we show another
method (<code>Harmony</code>+<code>Louvain</code>) to perform embedding
alignment and clustering.</p>
<p>First, we use the function <code>RunHarmonyLouvain()</code> in
<code>ProFAST</code> R package to select the number of clusters</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">RunHarmonyLouvain</span>(PRECASTObj, <span class="at">resolution =</span> <span class="fl">0.3</span>)</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>ARI_vec_louvain <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span>M, <span class="cf">function</span>(r) mclust<span class="sc">::</span><span class="fu">adjustedRandIndex</span>(PRECASTObj<span class="sc">@</span>resList<span class="sc">$</span>Louvain<span class="sc">$</span>cluster[[r]], yList[[r]]))</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="fu">print</span>(ARI_vec_louvain)</span></code></pre></div>
<p>Then, we use the function <code>RuniSCMEB()</code> in
<code>ProFAST</code> R package to jointly perform embedding alignment
and spatial clustering. The results are saved in the slot
<code>PRECASTObj@resList$iSCMEB</code>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">RuniSCMEB</span>(PRECASTObj, <span class="at">seed=</span><span class="dv">1</span>)</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="fu">str</span>(PRECASTObj<span class="sc">@</span>resList<span class="sc">$</span>iSCMEB)</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span>M, <span class="cf">function</span>(r) mclust<span class="sc">::</span><span class="fu">adjustedRandIndex</span>(PRECASTObj<span class="sc">@</span>resList<span class="sc">$</span>iSCMEB<span class="sc">$</span>cluster[[r]], yList[[r]]))</span></code></pre></div>
</div>
<div id="remove-unwanted-variations-in-the-log-normalized-expression-matrices" class="section level2">
<h2>Remove unwanted variations in the log-normalized expression
matrices</h2>
<p>In the following, we remove the unwanted variations in the
log-normalized expression matrices to obtain a combined log-normalized
expression matrix in a Seurat object. For this real data, we use
housekeeping genes as negative control to remove the unwanted
variations. Specifically, we leverage the batch effect embeddings
estimated through housekeeping gene expression matrix to capture
unwanted variations. Additionally, we utilize the cluster labels
obtained via iSC-MEB to retain the desired biological effects. The
posterior probability of cluster labels (RList) are in the slot
<code>PRECASTObj@resList$iSCMEB</code>.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">str</span>(PRECASTObj<span class="sc">@</span>resList<span class="sc">$</span>iSCMEB)</span></code></pre></div>
<p>To capture the unwanted variation, we first select the select the
housekeeping genes using the function <code>SelectHKgenes()</code>.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="do">## select the HK genes</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>HKgenes <span class="ot">&lt;-</span> <span class="fu">SelectHKgenes</span>(seuList, <span class="at">species=</span> <span class="st">&quot;Human&quot;</span>, <span class="at">HK.number=</span><span class="dv">200</span>)</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>seulist_HK <span class="ot">&lt;-</span> <span class="fu">lapply</span>(seuList, <span class="cf">function</span>(x) x[HKgenes, ])</span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a>seulist_HK</span></code></pre></div>
<p>Then, we integrate the two sections by removing the unwanted
variations by using <code>seulist_HK=seulist_HK</code> and
<code>Method = &quot;iSC-MEB&quot;</code> in the function
<code>IntegrateSRTData()</code>. After obtaining <code>seuInt</code>, we
will see there are three embeddings: <code>FAST</code>,
<code>iscmeb</code> and <code>position</code>, in the slot
<code>seuInt@reductions</code>. <code>FAST</code> are the embeddings
obtained by FAST model fitting and are uncorrected embeddings that may
includes the unwanted effects (i.e., batch effects);
<code>iscmeb</code>are the aligned embeddings obtained by iSC-MEB
fitting; and <code>position</code> are the spatial coordinates.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>seuInt <span class="ot">&lt;-</span> <span class="fu">IntegrateSRTData</span>(PRECASTObj, <span class="at">seulist_HK=</span>seulist_HK, <span class="at">Method =</span> <span class="st">&quot;iSC-MEB&quot;</span>,</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>                           <span class="at">seuList_raw=</span><span class="cn">NULL</span>, <span class="at">covariates_use=</span><span class="cn">NULL</span>, <span class="at">verbose=</span><span class="cn">TRUE</span>)</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a>seuInt</span></code></pre></div>
</div>
<div id="visualization" class="section level2">
<h2>Visualization</h2>
<p>First, user can choose a beautiful color schema using
<code>chooseColors()</code> in the R package <code>PRECAST</code>.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a>cols_cluster <span class="ot">&lt;-</span> <span class="fu">chooseColors</span>(<span class="at">palettes_name =</span> <span class="st">&quot;Nature 10&quot;</span>, <span class="at">n_colors =</span> <span class="dv">8</span>, <span class="at">plot_colors =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Then, we plot the spatial scatter plot for clusters using the
function <code>SpaPlot()</code> in the R package
<code>PRECAST</code>.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>p12 <span class="ot">&lt;-</span> <span class="fu">SpaPlot</span>(seuInt, <span class="at">item =</span> <span class="st">&quot;cluster&quot;</span>, <span class="at">batch =</span> <span class="cn">NULL</span>, <span class="at">point_size =</span> <span class="dv">1</span>, <span class="at">cols =</span> cols_cluster, <span class="at">combine =</span> <span class="cn">FALSE</span>)</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>p12 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(p12, <span class="cf">function</span>(x) x<span class="sc">+</span> <span class="fu">coord_flip</span>() <span class="sc">+</span> <span class="fu">scale_x_reverse</span>())</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="fu">drawFigs</span>(p12, <span class="at">layout.dim =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">common.legend =</span> T)</span></code></pre></div>
<p>We use the function <code>AddUMAP()</code> in the R package
<code>PRECAST</code> to obtain the three-dimensional components of UMAP
using the aligned embeddings in the reduction <code>harmony</code>.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a>seuInt <span class="ot">&lt;-</span> <span class="fu">AddUMAP</span>(seuInt, <span class="at">n_comp=</span><span class="dv">3</span>, <span class="at">reduction =</span> <span class="st">&#39;iscmeb&#39;</span>, <span class="at">assay =</span> <span class="st">&#39;RNA&#39;</span>)</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>seuInt</span></code></pre></div>
<p>We plot the spatial tNSE RGB plot to illustrate the performance in
extracting features.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>p13 <span class="ot">&lt;-</span> <span class="fu">SpaPlot</span>(seuInt, <span class="at">batch =</span> <span class="cn">NULL</span>, <span class="at">item =</span> <span class="st">&quot;RGB_UMAP&quot;</span>, <span class="at">point_size =</span> <span class="dv">1</span>, <span class="at">combine =</span> <span class="cn">FALSE</span>, <span class="at">text_size =</span> <span class="dv">15</span>)</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>p13 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(p13, <span class="cf">function</span>(x) x<span class="sc">+</span> <span class="fu">coord_flip</span>() <span class="sc">+</span> <span class="fu">scale_x_reverse</span>())</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a><span class="fu">drawFigs</span>(p13, <span class="at">layout.dim =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">common.legend =</span> <span class="cn">TRUE</span>, <span class="at">legend.position =</span> <span class="st">&quot;right&quot;</span>, <span class="at">align =</span> <span class="st">&quot;hv&quot;</span>)</span></code></pre></div>
<p>We use the function <code>AddUTSNE()</code> in the R package
<code>PRECAST</code> to obtain the two-dimensional components of UMAP
using the aligned embeddings in the reduction <code>iscmeb</code>, and
plot the tSNE plot based on the extracted features to check the
performance of integration.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a>seuInt <span class="ot">&lt;-</span> <span class="fu">AddTSNE</span>(seuInt, <span class="at">n_comp =</span> <span class="dv">2</span>, <span class="at">reduction =</span> <span class="st">&#39;iscmeb&#39;</span>, <span class="at">assay =</span> <span class="st">&#39;RNA&#39;</span>)</span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">dimPlot</span>(seuInt, <span class="at">item =</span> <span class="st">&quot;cluster&quot;</span>, <span class="at">point_size =</span> <span class="fl">0.5</span>, <span class="at">font_family =</span> <span class="st">&quot;serif&quot;</span>, <span class="at">cols =</span> cols_cluster,</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a>    <span class="at">border_col =</span> <span class="st">&quot;gray10&quot;</span>,  <span class="at">legend_pos =</span> <span class="st">&quot;right&quot;</span>)  <span class="co"># Times New Roman</span></span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">dimPlot</span>(seuInt, <span class="at">item =</span> <span class="st">&quot;batch&quot;</span>, <span class="at">point_size =</span> <span class="fl">0.5</span>, <span class="at">font_family =</span> <span class="st">&quot;serif&quot;</span>, <span class="at">legend_pos =</span> <span class="st">&quot;right&quot;</span>)</span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a></span>
<span id="cb22-6"><a href="#cb22-6" tabindex="-1"></a><span class="fu">drawFigs</span>(<span class="fu">list</span>(p1, p2), <span class="at">layout.dim =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">legend.position =</span> <span class="st">&quot;right&quot;</span>, <span class="at">align =</span> <span class="st">&quot;hv&quot;</span>)</span></code></pre></div>
</div>
<div id="combined-differential-epxression-analysis" class="section level2">
<h2>Combined differential epxression analysis</h2>
<p>Finally, we condut the combined differential expression analysis
using the integrated log-normalized expression matrix saved in the
<code>seuInt</code> object. The function <code>FindAllMarkers()</code>
in the <code>Seurat</code> R package is ued to achieve this
analysis.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a>dat_deg <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(seuInt)</span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>dat_deg <span class="sc">%&gt;%</span></span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a>    <span class="fu">top_n</span>(<span class="at">n =</span> n, <span class="at">wt =</span> avg_log2FC) <span class="ot">-&gt;</span> top10</span></code></pre></div>
<p>Plot dot plot of normalized expressions of the layer-specified marker
genes for each spatial domain identified by using the FAST
embeddings.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a>col_here <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;#F2E6AB&quot;</span>, <span class="st">&quot;#9C0141&quot;</span>) </span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a>marker_Seurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;151507_DEGmarkerTop_5.rds&quot;</span>)</span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a>marker_Seurat <span class="ot">&lt;-</span> <span class="fu">lapply</span>(marker_Seurat, <span class="cf">function</span>(x) <span class="fu">transferGeneNames</span>(x, <span class="at">Method=</span><span class="st">&#39;eg.db&#39;</span>))</span>
<span id="cb24-5"><a href="#cb24-5" tabindex="-1"></a>maker_genes <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(marker_Seurat, toupper))</span>
<span id="cb24-6"><a href="#cb24-6" tabindex="-1"></a>features <span class="ot">&lt;-</span> maker_genes[<span class="sc">!</span><span class="fu">duplicated</span>(maker_genes)]</span>
<span id="cb24-7"><a href="#cb24-7" tabindex="-1"></a>p_dot <span class="ot">&lt;-</span> <span class="fu">DotPlot</span>(seuInt, <span class="at">features=</span><span class="fu">unname</span>(features), <span class="at">cols=</span>col_here, <span class="co">#  idents = ident_here,</span></span>
<span id="cb24-8"><a href="#cb24-8" tabindex="-1"></a>              <span class="at">col.min =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">col.max =</span> <span class="dv">1</span>) <span class="sc">+</span> <span class="fu">coord_flip</span>()<span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">&quot;italic&quot;</span>))<span class="sc">+</span></span>
<span id="cb24-9"><a href="#cb24-9" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Domain&quot;</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="cn">NULL</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">12</span>, <span class="at">angle =</span> <span class="dv">25</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">family=</span><span class="st">&#39;serif&#39;</span>),</span>
<span id="cb24-10"><a href="#cb24-10" tabindex="-1"></a>                                      <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">12</span>, <span class="at">face=</span> <span class="st">&quot;italic&quot;</span>, <span class="at">family=</span><span class="st">&#39;serif&#39;</span>))</span>
<span id="cb24-11"><a href="#cb24-11" tabindex="-1"></a>p_dot</span>
<span id="cb24-12"><a href="#cb24-12" tabindex="-1"></a><span class="co"># table(unlist(yList), Idents(seuInt))</span></span></code></pre></div>
<p>Based on the marker genes, we annotate the Domains 1-9 as Layer5/6,
Layer3/4, Layer5, Layer2/3, Layer2/3, Layer6, WM, Layer5, Layer1. Then,
we rename the Domains using the function
<code>RenameIdents()</code>.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a>seuInt2 <span class="ot">&lt;-</span> <span class="fu">RenameIdents</span>(seuInt, <span class="st">&#39;1&#39;</span> <span class="ot">=</span> <span class="st">&#39;Layer6&#39;</span>, <span class="st">&#39;2&#39;</span> <span class="ot">=</span> <span class="st">&#39;Layer2/3&#39;</span>, <span class="st">&#39;3&#39;</span><span class="ot">=</span><span class="st">&quot;Layer5&quot;</span>, <span class="st">&#39;4&#39;</span><span class="ot">=</span><span class="st">&quot;WM&quot;</span>,</span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a>                           <span class="st">&#39;5&#39;</span><span class="ot">=</span><span class="st">&#39;Layer3/4&#39;</span>, <span class="st">&#39;6&#39;</span><span class="ot">=</span> <span class="st">&#39;Layer1&#39;</span>)</span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a><span class="fu">levels</span>(seuInt2)  <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Layer1&quot;</span>, <span class="st">&quot;Layer2/3&quot;</span>, <span class="st">&quot;Layer3/4&quot;</span>, <span class="st">&quot;Layer5&quot;</span>, <span class="st">&quot;Layer6&quot;</span>, <span class="st">&quot;WM&quot;</span>)</span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a>p_dot2 <span class="ot">&lt;-</span> <span class="fu">DotPlot</span>(seuInt2, <span class="at">features=</span><span class="fu">unname</span>(features), <span class="at">cols=</span>col_here, <span class="co">#  idents = ident_here,</span></span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a>              <span class="at">col.min =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">col.max =</span> <span class="dv">1</span>) <span class="sc">+</span> <span class="fu">coord_flip</span>()<span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">&quot;italic&quot;</span>))<span class="sc">+</span></span>
<span id="cb25-6"><a href="#cb25-6" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Domain&quot;</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="cn">NULL</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">12</span>, <span class="at">angle =</span> <span class="dv">25</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">family=</span><span class="st">&#39;serif&#39;</span>),</span>
<span id="cb25-7"><a href="#cb25-7" tabindex="-1"></a>                                      <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">12</span>, <span class="at">face=</span> <span class="st">&quot;italic&quot;</span>, <span class="at">family=</span><span class="st">&#39;serif&#39;</span>))</span>
<span id="cb25-8"><a href="#cb25-8" tabindex="-1"></a>p_dot2</span></code></pre></div>
<details>
<summary>
<strong>Session Info</strong>
</summary>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a><span class="co">#&gt; R version 4.4.1 (2024-06-14 ucrt)</span></span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64</span></span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a><span class="co">#&gt; Running under: Windows 11 x64 (build 26100)</span></span>
<span id="cb26-5"><a href="#cb26-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb26-6"><a href="#cb26-6" tabindex="-1"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb26-7"><a href="#cb26-7" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb26-8"><a href="#cb26-8" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb26-9"><a href="#cb26-9" tabindex="-1"></a><span class="co">#&gt; locale:</span></span>
<span id="cb26-10"><a href="#cb26-10" tabindex="-1"></a><span class="co">#&gt; [1] LC_COLLATE=C                               </span></span>
<span id="cb26-11"><a href="#cb26-11" tabindex="-1"></a><span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.utf8   </span></span>
<span id="cb26-12"><a href="#cb26-12" tabindex="-1"></a><span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.utf8</span></span>
<span id="cb26-13"><a href="#cb26-13" tabindex="-1"></a><span class="co">#&gt; [4] LC_NUMERIC=C                               </span></span>
<span id="cb26-14"><a href="#cb26-14" tabindex="-1"></a><span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.utf8    </span></span>
<span id="cb26-15"><a href="#cb26-15" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb26-16"><a href="#cb26-16" tabindex="-1"></a><span class="co">#&gt; time zone: Asia/Shanghai</span></span>
<span id="cb26-17"><a href="#cb26-17" tabindex="-1"></a><span class="co">#&gt; tzcode source: internal</span></span>
<span id="cb26-18"><a href="#cb26-18" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb26-19"><a href="#cb26-19" tabindex="-1"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb26-20"><a href="#cb26-20" tabindex="-1"></a><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span id="cb26-21"><a href="#cb26-21" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb26-22"><a href="#cb26-22" tabindex="-1"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb26-23"><a href="#cb26-23" tabindex="-1"></a><span class="co">#&gt;  [1] digest_0.6.37     R6_2.5.1          fastmap_1.2.0     xfun_0.47        </span></span>
<span id="cb26-24"><a href="#cb26-24" tabindex="-1"></a><span class="co">#&gt;  [5] cachem_1.1.0      knitr_1.48        htmltools_0.5.8.1 rmarkdown_2.28   </span></span>
<span id="cb26-25"><a href="#cb26-25" tabindex="-1"></a><span class="co">#&gt;  [9] lifecycle_1.0.4   cli_3.6.3         sass_0.4.9        jquerylib_0.1.4  </span></span>
<span id="cb26-26"><a href="#cb26-26" tabindex="-1"></a><span class="co">#&gt; [13] compiler_4.4.1    rstudioapi_0.16.0 tools_4.4.1       evaluate_1.0.0   </span></span>
<span id="cb26-27"><a href="#cb26-27" tabindex="-1"></a><span class="co">#&gt; [17] bslib_0.8.0       yaml_2.3.10       rlang_1.1.4       jsonlite_1.8.9</span></span></code></pre></div>
</details>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
