<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Wei Liu" />

<meta name="date" content="2025-12-14" />

<title>FAST: single DLPFC section</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">FAST: single DLPFC section</h1>
<h4 class="author">Wei Liu</h4>
<h4 class="date">2025-12-14</h4>



<p>This vignette introduces the FAST workflow for the analysis of
single-section rather than multi-secitons data, one humn dorsolateral
prefrontal cortex (DLPFC) spatial transcriptomics dataset. In this
vignette, the workflow of FAST consists of three steps</p>
<ul>
<li>Independent preprocessing</li>
<li>Spatial dimension reduction using FAST</li>
<li>Downstream analysis (i.e. , embedding alignment and spatial
clustering using iSC-MEB, visualization of clusters and embeddings,
remove unwanted variation, differential expression analysis)</li>
</ul>
<p>We demonstrate the use of FAST to one DLPFC Visium data that are <a href="https://github.com/feiyoung/ProFAST/tree/main/vignettes_data">here</a>,
which can be downloaded to the current working path by the following
command:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>githubURL <span class="ot">&lt;-</span> <span class="st">&quot;https://github.com/feiyoung/FAST/blob/main/vignettes_data/seulist2_ID9_10.RDS?raw=true&quot;</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">download.file</span>(githubURL,<span class="st">&quot;seulist2_ID9_10.RDS&quot;</span>,<span class="at">mode=</span><span class="st">&#39;wb&#39;</span>)</span></code></pre></div>
<p>Then load to R. Here, we only focus one section.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>dlpfc2 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;./seulist2_ID9_10.RDS&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>dlpfc <span class="ot">&lt;-</span> dlpfc2[[<span class="dv">1</span>]]</span></code></pre></div>
<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">library</span>(ProFAST) <span class="co"># load the package of FAST method</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">library</span>(PRECAST)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span></code></pre></div>
<div id="view-the-dlpfc-data" class="section level2">
<h2>View the DLPFC data</h2>
<p>First, we view the the spatial transcriptomics data with Visium
platform. There are ~15000 genes and ~3600 spots.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>dlpfc <span class="do">## a list including three Seurat object with default assay: RNA</span></span></code></pre></div>
<p>We observed that the genes are Ensembl IDs. In the following, we will
transfer the Ensembl IDs to gene symbols for matching the housekeeping
genes in the downstream analysis for removing the unwanted
variations.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">row.names</span>(dlpfc)[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>count <span class="ot">&lt;-</span> dlpfc[[<span class="st">&#39;RNA&#39;</span>]]<span class="sc">@</span>counts</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="fu">row.names</span>(count) <span class="ot">&lt;-</span> <span class="fu">unname</span>(<span class="fu">transferGeneNames</span>(<span class="fu">row.names</span>(count), <span class="at">now_name =</span> <span class="st">&quot;ensembl&quot;</span>,</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>                                                 <span class="at">to_name=</span><span class="st">&quot;symbol&quot;</span>,</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>                                                 <span class="at">species=</span><span class="st">&quot;Human&quot;</span>, <span class="at">Method=</span><span class="st">&#39;eg.db&#39;</span>))</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">row.names</span>(count)[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> count, <span class="at">meta.data =</span> dlpfc<span class="sc">@</span>meta.data)</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>seu</span></code></pre></div>
</div>
<div id="preprocessing" class="section level2">
<h2>Preprocessing</h2>
<p>We show how to preprocessing before fitting FAST, including
log-normalization (if user use the gaussian version of FAST), and select
highly variable genes.</p>
<ul>
<li>Note: the spatial coordinates must be contained in the meta data and
named as <code>row</code> and <code>col</code>, which benefits the
identification of spaital coordinates by FAST.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="do">## Check the spatial coordinates: they are named as &quot;row&quot; and &quot;col&quot;!</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">head</span>(seu<span class="sc">@</span>meta.data))</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(seu)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(seu)</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="fu">print</span>(seu)</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="fu">print</span>(seu[[<span class="st">&#39;RNA&#39;</span>]]<span class="sc">@</span>var.features[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span></code></pre></div>
<details>
<summary>
<strong>Find spatially variable genes</strong>
</summary>
<p>Users can also use the spatially variable genes by the following
command:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2023</span>)</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>seu <span class="ot">&lt;-</span> DR.SC<span class="sc">::</span><span class="fu">FindSVGs</span>(seu)</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="do">### Check the results</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="fu">print</span>(seu[[<span class="st">&#39;RNA&#39;</span>]]<span class="sc">@</span>var.features[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span></code></pre></div>
</details>
</div>
<div id="fit-fast-for-this-data" class="section level2">
<h2>Fit FAST for this data</h2>
<p>For function <code>FAST_single</code>, users can specify the number
of factors <code>q</code> and the fitted model <code>fit.model</code>.
The <code>q</code> sets the number of spatial factors to be extracted,
and a lareger one means more information to be extracted but higher
computaional cost. The <code>fit.model</code> specifies the version of
FAST to be fitted. The Gaussian version (<code>gaussian</code>) models
the log-normalized matrix while the Poisson verion
(<code>poisson</code>) models the count matrix; default as
<code>poisson</code>. (Note: The computational time required to run the
analysis on personal PCs is approximately ~0.5 minute on a personal
PC.)</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>Adj_sp <span class="ot">&lt;-</span> <span class="fu">AddAdj</span>(<span class="fu">as.matrix</span>(seu<span class="sc">@</span>meta.data[,<span class="fu">c</span>(<span class="st">&quot;row&quot;</span>, <span class="st">&quot;col&quot;</span>)]), <span class="at">platform =</span> <span class="st">&quot;Visium&quot;</span>)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="do">### set q= 15 here</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2023</span>)</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">FAST_single</span>(seu, <span class="at">Adj_sp=</span>Adj_sp, <span class="at">q=</span> <span class="dv">15</span>, <span class="at">fit.model=</span><span class="st">&#39;poisson&#39;</span>)</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>seu</span></code></pre></div>
<details>
<summary>
<strong>Run possion version</strong>
</summary>
<p>Users can also use the gaussian version by the following command:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>Adj_sp <span class="ot">&lt;-</span> <span class="fu">AddAdj</span>(<span class="fu">as.matrix</span>(seu<span class="sc">@</span>meta.data[,<span class="fu">c</span>(<span class="st">&quot;row&quot;</span>, <span class="st">&quot;col&quot;</span>)]), <span class="at">platform =</span> <span class="st">&quot;Visium&quot;</span>)</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2023</span>)</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">FAST_single</span>(dlpfc, <span class="at">Adj_sp=</span>Adj_sp, <span class="at">q=</span> <span class="dv">15</span>, <span class="at">fit.model=</span><span class="st">&#39;gaussian&#39;</span>)</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="do">### Check the results</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>seu</span></code></pre></div>
</details>
<div id="evaluate-the-adjusted-mcfaddens-pseudo-r-square" class="section level3">
<h3>Evaluate the adjusted McFadden’s pseudo R-square</h3>
<p>Next, we investigate the performance of dimension reduction by
calculating the adjusted McFadden’s pseudo R-square. The manual
annotations are regarded as the ground truth in the
<code>meta.data</code> of <code>seu</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="do">## Obtain the true labels</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>y <span class="ot">&lt;-</span> seu<span class="sc">$</span>layer_guess_reordered</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="do">### Evaluate the MacR2</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>Mac <span class="ot">&lt;-</span>  <span class="fu">get_r2_mcfadden</span>(<span class="fu">Embeddings</span>(seu, <span class="at">reduction=</span><span class="st">&#39;fast&#39;</span>), y)</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="do">### output them</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;MacFadden&#39;s R-square of FAST is &quot;</span>, <span class="fu">round</span>(Mac, <span class="dv">3</span>)))</span></code></pre></div>
</div>
</div>
<div id="clustering-analysis-based-on-fast-embedding" class="section level2">
<h2>Clustering analysis based on FAST embedding</h2>
<p>Based on the embeddings from FAST, we use <code>Louvain</code> to
perform clustering. In this downstream analysis, other methods for
clustering can be also used.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(seu, <span class="at">reduction =</span> <span class="st">&#39;fast&#39;</span>)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(seu, <span class="at">resolution =</span> <span class="fl">0.4</span>)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>seu<span class="sc">$</span>fast.cluster <span class="ot">&lt;-</span> seu<span class="sc">$</span>seurat_clusters</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>ARI.fast <span class="ot">&lt;-</span> mclust<span class="sc">::</span><span class="fu">adjustedRandIndex</span>(y, seu<span class="sc">$</span>fast.cluster)</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;ARI of PCA is &quot;</span>, <span class="fu">round</span>(ARI.fast, <span class="dv">3</span>)))</span></code></pre></div>
<p>For comparison, we also run PCA to obtain PCA embeddings, and then
conduct louvain clustering.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(seu)</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(seu, <span class="at">npcs=</span><span class="dv">15</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>Mac.pca <span class="ot">&lt;-</span> <span class="fu">get_r2_mcfadden</span>(<span class="fu">Embeddings</span>(seu, <span class="at">reduction=</span><span class="st">&#39;pca&#39;</span>), y)</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;MacFadden&#39;s R-square of PCA is &quot;</span>, <span class="fu">round</span>(Mac.pca, <span class="dv">3</span>)))</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(seu, <span class="at">reduction =</span> <span class="st">&#39;pca&#39;</span>, <span class="at">graph.name =</span><span class="st">&quot;pca.graph&quot;</span>)</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(seu, <span class="at">resolution =</span> <span class="fl">0.8</span>,<span class="at">graph.name =</span> <span class="st">&#39;pca.graph&#39;</span>) </span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>seu<span class="sc">$</span>pca.cluster <span class="ot">&lt;-</span> seu<span class="sc">$</span>seurat_clusters</span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a>ARI.pca <span class="ot">&lt;-</span> mclust<span class="sc">::</span><span class="fu">adjustedRandIndex</span>(y, seu<span class="sc">$</span>pca.cluster)</span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;ARI of PCA is &quot;</span>, <span class="fu">round</span>(ARI.pca, <span class="dv">3</span>)))</span></code></pre></div>
</div>
<div id="visualization" class="section level2">
<h2>Visualization</h2>
<p>First, user can choose a beautiful color schema using
<code>chooseColors()</code> in the R package <code>PRECAST</code>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>cols_cluster <span class="ot">&lt;-</span> <span class="fu">chooseColors</span>(<span class="at">palettes_name =</span> <span class="st">&quot;Nature 10&quot;</span>, <span class="at">n_colors =</span> <span class="dv">8</span>, <span class="at">plot_colors =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Then, we plot the spatial scatter plot for clusters using the
function <code>DimPlot()</code> in the R package <code>Seurat</code>. We
observe that the clusters from PCA are more messy while the clusters
from FAST are more smoothing in spatial coordinates.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>seu <span class="ot">&lt;-</span> PRECAST<span class="sc">::</span><span class="fu">Add_embed</span>(<span class="at">embed =</span> <span class="fu">as.matrix</span>(seu<span class="sc">@</span>meta.data[,<span class="fu">c</span>(<span class="st">&quot;row&quot;</span>, <span class="st">&quot;col&quot;</span>)]), seu, <span class="at">embed_name =</span> <span class="st">&#39;Spatial&#39;</span>)</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>seu</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seu, <span class="at">reduction =</span> <span class="st">&#39;Spatial&#39;</span>, <span class="at">group.by =</span> <span class="st">&#39;pca.cluster&#39;</span>,<span class="at">cols =</span> cols_cluster, <span class="at">pt.size =</span> <span class="fl">1.5</span>)</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seu, <span class="at">reduction =</span> <span class="st">&#39;Spatial&#39;</span>, <span class="at">group.by =</span> <span class="st">&#39;fast.cluster&#39;</span>,<span class="at">cols =</span> cols_cluster, <span class="at">pt.size =</span> <span class="fl">1.5</span>)</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="fu">drawFigs</span>(<span class="fu">list</span>(p1, p2),<span class="at">layout.dim =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>) )</span></code></pre></div>
<p>Next, we visualize the clusters from <code>FAST</code> on the UMAP
space, and observe the clusters are well separated in general.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(seu, <span class="at">reduction =</span> <span class="st">&quot;fast&quot;</span>, <span class="at">dims=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>)</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>seu</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="fu">DimPlot</span>(seu, <span class="at">reduction=</span><span class="st">&#39;umap&#39;</span>, <span class="at">group.by =</span> <span class="st">&quot;fast.cluster&quot;</span>)</span></code></pre></div>
</div>
<div id="differential-epxression-analysis" class="section level2">
<h2>Differential epxression analysis</h2>
<p>Finally, we condut the differential expression (DE) analysis. The
function <code>FindAllMarkers()</code> in the <code>Seurat</code> R
package is ued to achieve this analysis. And we extract the top five DE
genes.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">Idents</span>(seu) <span class="ot">&lt;-</span> seu<span class="sc">$</span>fast.cluster</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>dat_deg <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(seu)</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a>dat_deg <span class="sc">%&gt;%</span></span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a>    <span class="fu">top_n</span>(<span class="at">n =</span> n, <span class="at">wt =</span> avg_log2FC) <span class="ot">-&gt;</span> top5</span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a>top5</span></code></pre></div>
<details>
<summary>
<strong>Session Info</strong>
</summary>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a><span class="co">#&gt; R version 4.4.1 (2024-06-14 ucrt)</span></span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64</span></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="co">#&gt; Running under: Windows 11 x64 (build 26100)</span></span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a><span class="co">#&gt; locale:</span></span>
<span id="cb18-10"><a href="#cb18-10" tabindex="-1"></a><span class="co">#&gt; [1] LC_COLLATE=C                               </span></span>
<span id="cb18-11"><a href="#cb18-11" tabindex="-1"></a><span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.utf8   </span></span>
<span id="cb18-12"><a href="#cb18-12" tabindex="-1"></a><span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.utf8</span></span>
<span id="cb18-13"><a href="#cb18-13" tabindex="-1"></a><span class="co">#&gt; [4] LC_NUMERIC=C                               </span></span>
<span id="cb18-14"><a href="#cb18-14" tabindex="-1"></a><span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.utf8    </span></span>
<span id="cb18-15"><a href="#cb18-15" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb18-16"><a href="#cb18-16" tabindex="-1"></a><span class="co">#&gt; time zone: Asia/Shanghai</span></span>
<span id="cb18-17"><a href="#cb18-17" tabindex="-1"></a><span class="co">#&gt; tzcode source: internal</span></span>
<span id="cb18-18"><a href="#cb18-18" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb18-19"><a href="#cb18-19" tabindex="-1"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb18-20"><a href="#cb18-20" tabindex="-1"></a><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span id="cb18-21"><a href="#cb18-21" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb18-22"><a href="#cb18-22" tabindex="-1"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb18-23"><a href="#cb18-23" tabindex="-1"></a><span class="co">#&gt;  [1] digest_0.6.37     R6_2.5.1          fastmap_1.2.0     xfun_0.47        </span></span>
<span id="cb18-24"><a href="#cb18-24" tabindex="-1"></a><span class="co">#&gt;  [5] cachem_1.1.0      knitr_1.48        htmltools_0.5.8.1 rmarkdown_2.28   </span></span>
<span id="cb18-25"><a href="#cb18-25" tabindex="-1"></a><span class="co">#&gt;  [9] lifecycle_1.0.4   cli_3.6.3         sass_0.4.9        jquerylib_0.1.4  </span></span>
<span id="cb18-26"><a href="#cb18-26" tabindex="-1"></a><span class="co">#&gt; [13] compiler_4.4.1    rstudioapi_0.16.0 tools_4.4.1       evaluate_1.0.0   </span></span>
<span id="cb18-27"><a href="#cb18-27" tabindex="-1"></a><span class="co">#&gt; [17] bslib_0.8.0       yaml_2.3.10       rlang_1.1.4       jsonlite_1.8.9</span></span></code></pre></div>
</details>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
