<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Wei Liu" />

<meta name="date" content="2025-03-27" />

<title>CoFAST: PBMC scRNA-seq data coembedding</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">CoFAST: PBMC scRNA-seq data
coembedding</h1>
<h4 class="author">Wei Liu</h4>
<h4 class="date">2025-03-27</h4>



<p>This vignette introduces the CoFAST workflow for the analysis of
PBMC3k single-cell RNA sequencing dataset. In this vignette, the
workflow of CoFAST consists of three steps</p>
<ul>
<li>Independent preprocessing and model setting</li>
<li>Coembedding dimension reduction</li>
<li>Downstream analysis (i.e.Â , signature gene analysis, visualization
of cell types and coembeddings)</li>
</ul>
<div id="load-and-view-data" class="section level2">
<h2>Load and view data</h2>
<p>We demonstrate the use of CoFAST to PBMC3k data that are in the
SeuratData package, which can be downloaded to the current working path
by the following command:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2024</span>) <span class="co"># set a random seed for reproducibility.</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>pbmc3k <span class="ot">&lt;-</span> SeuratData<span class="sc">::</span><span class="fu">LoadData</span>(<span class="st">&quot;pbmc3k&quot;</span>)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="do">## filter the seurat_annotation is NA</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(pbmc3k<span class="sc">$</span>seurat_annotations))</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>pbmc3k <span class="ot">&lt;-</span> pbmc3k[,idx]</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>pbmc3k</span></code></pre></div>
<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">library</span>(ProFAST) <span class="co"># load the package of FAST method</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span></code></pre></div>
</div>
<div id="preprocessing" class="section level2">
<h2>Preprocessing</h2>
<p>First, we normalize the data.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>pbmc3k <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(pbmc3k)</span></code></pre></div>
<p>Then, we select the variable genes.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>pbmc3k <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(pbmc3k)</span></code></pre></div>
</div>
<div id="coembedding-using-non-centered-factor-model" class="section level2">
<h2>Coembedding using non-centered factor model</h2>
<p>We introduce how to use the non-centered factor model (NCFM) to
perform coembedding for this scRNA-seq data. First, we determine the
dimension of coembeddings. Here, we use the parallel analysis method to
select the dimension.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>dat_cor <span class="ot">&lt;-</span> <span class="fu">diagnostic.cor.eigs</span>(pbmc3k)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>q_est <span class="ot">&lt;-</span> <span class="fu">attr</span>(dat_cor, <span class="st">&quot;q_est&quot;</span>)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;q_est = &quot;</span>, q_est, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span></code></pre></div>
<p>Subsequently, we calculate coembeddings by utilizing NCFM, and
observe that the <code>reductions</code> field acquires an additional
component named <code>ncfm</code>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>pbmc3k <span class="ot">&lt;-</span> <span class="fu">NCFM</span>(pbmc3k, <span class="at">q =</span> q_est)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>pbmc3k</span></code></pre></div>
</div>
<div id="downstream-analysis" class="section level2">
<h2>Downstream analysis</h2>
<p>In the following, we show how to find the signature genes based on
comebeddings. First, we calculate the distance matrix.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>pbmc3k <span class="ot">&lt;-</span> <span class="fu">pdistance</span>(pbmc3k, <span class="at">reduction =</span> <span class="st">&quot;ncfm&quot;</span>)</span></code></pre></div>
<p>Next, we find the signature genes for each cell type</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">table</span>(pbmc3k<span class="sc">$</span>seurat_annotations))</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="fu">Idents</span>(pbmc3k) <span class="ot">&lt;-</span> pbmc3k<span class="sc">$</span>seurat_annotations</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>df_sig_list <span class="ot">&lt;-</span> <span class="fu">find.signature.genes</span>(pbmc3k)</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="fu">str</span>(df_sig_list)</span></code></pre></div>
<p>Then, we obtain the top five signature genes and organize them into a
data.frame. The colname <code>distance</code> means the distance between
gene (i.e., VPREB3) and cells with the specific cell type (i.e., B
cell), which is calculated based on the coembedding of genes and cells
in the coembedding space. The distance is smaller, the association
between gene and the cell type is stronger. The colname
<code>expr.prop</code> represents the expression proportion of the gene
(i.e., VPREB3) within the cell type (i.e., B cell). The colname
<code>label</code> means the cell types and colname <code>gene</code>
denotes the gene name. By the data.frame object, we know
<code>VPREB3</code> is the one of the top signature gene of B cell.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">get.top.signature.dat</span>(df_sig_list, <span class="at">ntop =</span> <span class="dv">5</span>, <span class="at">expr.prop.cutoff =</span> <span class="fl">0.1</span>)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">head</span>(dat)</span></code></pre></div>
<p>Next, we calculate the UMAP projections of coembeddings of cells and
the selected signature genes.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>pbmc3k <span class="ot">&lt;-</span> <span class="fu">coembedding_umap</span>(</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>  pbmc3k, <span class="at">reduction =</span> <span class="st">&quot;ncfm&quot;</span>, <span class="at">reduction.name =</span> <span class="st">&quot;UMAP&quot;</span>,</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>  <span class="at">gene.set =</span> <span class="fu">unique</span>(dat<span class="sc">$</span>gene))</span></code></pre></div>
<p>Furthermore, we visualize the cells and top five signature genes of B
cell in the UMAP space of coembedding. We observe that the UMAP
projections of the five signature genes are near to B cells, which
indicates these genes are enriched in B cells.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="do">## choose beutifual colors</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>cols_cluster <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;black&quot;</span>, PRECAST<span class="sc">::</span><span class="fu">chooseColors</span>(<span class="at">palettes_name =</span> <span class="st">&quot;Light 13&quot;</span>, <span class="at">n_colors =</span> <span class="dv">9</span>, <span class="at">plot_colors =</span> <span class="cn">TRUE</span>))</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">coembed_plot</span>(</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>   pbmc3k, <span class="at">reduction =</span> <span class="st">&quot;UMAP&quot;</span>,</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>   <span class="at">gene_txtdata =</span> <span class="fu">subset</span>(dat, label<span class="sc">==</span><span class="st">&#39;B&#39;</span>), </span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>   <span class="at">cols=</span>cols_cluster,<span class="at">pt_text_size =</span> <span class="dv">3</span>)</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>p1</span></code></pre></div>
<p>Then, we visualize the cells and top five signature genes of all
involved cell types in the UMAP space of coembedding. We observe that
the UMAP projections of the five signature genes are near to the
corresponding cell type, which indicates these genes are enriched in the
corresponding cells.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">coembed_plot</span>(</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>   pbmc3k, <span class="at">reduction =</span> <span class="st">&quot;UMAP&quot;</span>,</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>   <span class="at">gene_txtdata =</span> dat, <span class="at">cols=</span>cols_cluster,</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>   <span class="at">pt_text_size =</span> <span class="dv">3</span>)</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>p2</span></code></pre></div>
<p>In addtion, we can fully take advantages of the visualization
functions in <code>Seurat</code> package for visualization. The
following is an example that visualizes the cell types on the UMAP
space.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>cols_type <span class="ot">&lt;-</span> cols_cluster[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="fu">names</span>(cols_type)<span class="ot">&lt;-</span>  <span class="fu">sort</span>(<span class="fu">levels</span>(<span class="fu">Idents</span>(pbmc3k)))</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="fu">DimPlot</span>(pbmc3k, <span class="at">reduction =</span> <span class="st">&#39;UMAP&#39;</span>, <span class="at">cols=</span>cols_type)</span></code></pre></div>
<p>Then, there is another example that we plot the two signature genes
of B cell on UMAP space, in which we observed the high expression in B
cells in constrast to other cell types.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="fu">FeaturePlot</span>(pbmc3k, <span class="at">reduction =</span> <span class="st">&#39;UMAP&#39;</span>, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;CD79A&quot;</span>, <span class="st">&quot;VPREB3&quot;</span>))</span></code></pre></div>
<details>
<summary>
<strong>Session Info</strong>
</summary>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="co">#&gt; R version 4.4.1 (2024-06-14 ucrt)</span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="co">#&gt; Running under: Windows 11 x64 (build 26100)</span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a><span class="co">#&gt; locale:</span></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a><span class="co">#&gt; [1] LC_COLLATE=C                               </span></span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a><span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.utf8   </span></span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a><span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.utf8</span></span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a><span class="co">#&gt; [4] LC_NUMERIC=C                               </span></span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a><span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.utf8    </span></span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb15-16"><a href="#cb15-16" tabindex="-1"></a><span class="co">#&gt; time zone: Asia/Shanghai</span></span>
<span id="cb15-17"><a href="#cb15-17" tabindex="-1"></a><span class="co">#&gt; tzcode source: internal</span></span>
<span id="cb15-18"><a href="#cb15-18" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb15-19"><a href="#cb15-19" tabindex="-1"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb15-20"><a href="#cb15-20" tabindex="-1"></a><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span id="cb15-21"><a href="#cb15-21" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb15-22"><a href="#cb15-22" tabindex="-1"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb15-23"><a href="#cb15-23" tabindex="-1"></a><span class="co">#&gt;  [1] digest_0.6.37     R6_2.5.1          fastmap_1.2.0     xfun_0.47        </span></span>
<span id="cb15-24"><a href="#cb15-24" tabindex="-1"></a><span class="co">#&gt;  [5] cachem_1.1.0      knitr_1.48        htmltools_0.5.8.1 rmarkdown_2.28   </span></span>
<span id="cb15-25"><a href="#cb15-25" tabindex="-1"></a><span class="co">#&gt;  [9] lifecycle_1.0.4   cli_3.6.3         sass_0.4.9        jquerylib_0.1.4  </span></span>
<span id="cb15-26"><a href="#cb15-26" tabindex="-1"></a><span class="co">#&gt; [13] compiler_4.4.1    rstudioapi_0.16.0 tools_4.4.1       evaluate_1.0.0   </span></span>
<span id="cb15-27"><a href="#cb15-27" tabindex="-1"></a><span class="co">#&gt; [17] bslib_0.8.0       yaml_2.3.10       rlang_1.1.4       jsonlite_1.8.9</span></span></code></pre></div>
</details>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
